### PR TITLE
clamp (nano)second values to unproblematic range, fixes #2304

### LINKED ISSUE #2304: way-off fs timestamps crash msgpack
https://github.com/msgpack/msgpack-python/issues/221

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
filesystem -> clamp -> archive (create)

### CODE REVIEW

File: src/borg/helpers.py [hunk 1]
```diff
@@ -818,12 +818,34 @@ def SortBySpec(text):
     return text.replace('timestamp', 'ts')
 
 
+# Not too rarely, we get crappy timestamps from the fs, that overflow some computations.
+# As they are crap anyway, nothing is lost if we just clamp them to the max valid value.
+# msgpack can only pack uint64. datetime is limited to year 9999.
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: better clamp to (positive) max_int64?

when does python switch from simple into to long int?

Comment (enkore) [inline src/borg/helpers.py]: The internal representation is unified (it is always a variable-length integer); the item length is the native size of the platform.

Comment (ThomasWaldmann) [inline src/borg/helpers.py]: ok, but assuming we are on a 64bit platform, the native datatype being used is then int64, not uint64.

so, guess it is more efficient if we clamp to 0 .. 2**63 - 1?

**Comment by enkore** [comment]
> archive -> clamp -> filesystem (extract, maybe old broken archives?)

I'm not sure that this part is necessary; we couldn't store these large timestamps before, but filesystems have a large variance in the supported ranges. The os.utime call is already wrapped in OSError, so there's likely nothing needed here.

> archive -> clamp -> FUSE (mount, maybe old broken archives?)

I guess this depends on the size of the st_Xtime and st_Xtime_ns fields. At least the latter ought to be >= 64 bits, so no problems there. If the former can be 32 bits it may be problematic, but evidently llfuseis liberal here with types (or the code path has never been used) -- since we're passing `float`s to these fields.

File: src/borg/helpers.py (hunk 1 above)
Comment (enkore) [inline src/borg/helpers.py]: IIRC the sign is encoded separately from the value itself

**Comment by ThomasWaldmann** [comment]
@enkore ok. i'll remove the unneeded safe_X calls for fuse and extract.

**Comment by ThomasWaldmann** [comment]
assuming tests are ok: collapse, merge?

**Pull request merged**